File: specs/typespec/agency-domains/agency-domains.go
package agencydomains

import (
	"time"

	"vetchium-api-server.typespec/common"
)

// AgencyDomainVerificationToken - secret expected in DNS TXT record
type AgencyDomainVerificationToken string

// AgencyDomainVerificationStatus enum
type AgencyDomainVerificationStatus string

const (
	AgencyDomainVerificationStatusPending  AgencyDomainVerificationStatus = "PENDING"
	AgencyDomainVerificationStatusVerified AgencyDomainVerificationStatus = "VERIFIED"
	AgencyDomainVerificationStatusFailing  AgencyDomainVerificationStatus = "FAILING"
)

// Constants for agency domain verification
const (
	AgencyTokenExpiryDays             = 7
	AgencyVerificationIntervalDays    = 60
	AgencyGracePeriodDays             = 14
	AgencyMaxConsecutiveFailures      = 3
	AgencyVerificationCooldownMinutes = 60 // Rate limit: 1 hour between verification requests
)

// ============================================
// Agency Domain Verification Flow
// ============================================

type AgencyClaimDomainRequest struct {
	Domain common.DomainName `json:"domain"`
}

func (r AgencyClaimDomainRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type AgencyClaimDomainResponse struct {
	Domain            string                        `json:"domain"`
	VerificationToken AgencyDomainVerificationToken `json:"verification_token"`
	ExpiresAt         time.Time                     `json:"expires_at"`
	Instructions      string                        `json:"instructions"`
}

type AgencyVerifyDomainRequest struct {
	Domain common.DomainName `json:"domain"`
}

func (r AgencyVerifyDomainRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type AgencyVerifyDomainResponse struct {
	Status     AgencyDomainVerificationStatus `json:"status"`
	VerifiedAt *time.Time                     `json:"verified_at,omitempty"`
	Message    *string                        `json:"message,omitempty"`
}

type AgencyGetDomainStatusRequest struct {
	Domain common.DomainName `json:"domain"`
}

func (r AgencyGetDomainStatusRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError

	if r.Domain == "" {
		errs = append(errs, common.NewValidationError("domain", common.ErrRequired))
	} else if err := r.Domain.Validate(); err != nil {
		errs = append(errs, common.NewValidationError("domain", err))
	}

	return errs
}

type AgencyGetDomainStatusResponse struct {
	Domain                    string                         `json:"domain"`
	Status                    AgencyDomainVerificationStatus `json:"status"`
	VerificationToken         *AgencyDomainVerificationToken `json:"verification_token,omitempty"`
	ExpiresAt                 *time.Time                     `json:"expires_at,omitempty"`
	LastVerifiedAt            *time.Time                     `json:"last_verified_at,omitempty"`
	CanRequestVerification    bool                           `json:"can_request_verification"`
	LastAttemptedAt           *time.Time                     `json:"last_attempted_at,omitempty"`
	NextVerificationAllowedAt *time.Time                     `json:"next_verification_allowed_at,omitempty"`
}

type AgencyListDomainStatusRequest struct {
	PaginationKey *string `json:"pagination_key,omitempty"`
}

func (r AgencyListDomainStatusRequest) Validate() []common.ValidationError {
	return nil
}

type AgencyListDomainStatusItem struct {
	Domain                    string                         `json:"domain"`
	Status                    AgencyDomainVerificationStatus `json:"status"`
	VerificationToken         *AgencyDomainVerificationToken `json:"verification_token,omitempty"`
	ExpiresAt                 *time.Time                     `json:"expires_at,omitempty"`
	LastVerifiedAt            *time.Time                     `json:"last_verified_at,omitempty"`
	CanRequestVerification    bool                           `json:"can_request_verification"`
	LastAttemptedAt           *time.Time                     `json:"last_attempted_at,omitempty"`
	NextVerificationAllowedAt *time.Time                     `json:"next_verification_allowed_at,omitempty"`
}

type AgencyListDomainStatusResponse struct {
	Items             []AgencyListDomainStatusItem `json:"items"`
	NextPaginationKey *string                      `json:"next_pagination_key,omitempty"`
}

File: specs/typespec/agency-domains/agency-domains.ts
import {
	type DomainName,
	type ValidationError,
	newValidationError,
	validateDomainName,
	ERR_REQUIRED,
} from "../common/common";

// Domain Verification Token - secret expected in DNS TXT record
export type AgencyDomainVerificationToken = string;

// Domain verification status enum
export const AgencyDomainVerificationStatus = {
	PENDING: "PENDING",
	VERIFIED: "VERIFIED",
	FAILING: "FAILING",
} as const;

export type AgencyDomainVerificationStatus =
	(typeof AgencyDomainVerificationStatus)[keyof typeof AgencyDomainVerificationStatus];

// Constants for agency domain verification
export const AgencyTokenExpiryDays = 7;
export const AgencyVerificationIntervalDays = 60;
export const AgencyGracePeriodDays = 14;
export const AgencyMaxConsecutiveFailures = 3;
export const AgencyVerificationCooldownMinutes = 60; // Rate limit: 1 hour between verification requests

// ============================================
// Agency Domain Verification Flow
// ============================================

export interface AgencyClaimDomainRequest {
	domain: DomainName;
}

export function validateAgencyClaimDomainRequest(
	request: AgencyClaimDomainRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	return errs;
}

export interface AgencyClaimDomainResponse {
	domain: string;
	verification_token: AgencyDomainVerificationToken;
	expires_at: string; // ISO 8601 datetime
	instructions: string;
}

export interface AgencyVerifyDomainRequest {
	domain: DomainName;
}

export function validateAgencyVerifyDomainRequest(
	request: AgencyVerifyDomainRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	return errs;
}

export interface AgencyVerifyDomainResponse {
	status: AgencyDomainVerificationStatus;
	verified_at?: string; // ISO 8601 datetime
	message?: string;
}

export interface AgencyGetDomainStatusRequest {
	domain: DomainName;
}

export function validateAgencyGetDomainStatusRequest(
	request: AgencyGetDomainStatusRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	if (!request.domain) {
		errs.push(newValidationError("domain", ERR_REQUIRED));
	} else {
		const domainErr = validateDomainName(request.domain);
		if (domainErr) {
			errs.push(newValidationError("domain", domainErr));
		}
	}

	return errs;
}

export interface AgencyGetDomainStatusResponse {
	domain: string;
	status: AgencyDomainVerificationStatus;
	verification_token?: AgencyDomainVerificationToken;
	expires_at?: string; // ISO 8601 datetime
	last_verified_at?: string; // ISO 8601 datetime
	can_request_verification: boolean;
	last_attempted_at?: string; // ISO 8601 datetime — when verification was last requested
	next_verification_allowed_at?: string; // ISO 8601 datetime — earliest time next request is allowed (only when !can_request_verification)
}

export interface AgencyListDomainStatusRequest {
	pagination_key?: string;
}

export function validateAgencyListDomainStatusRequest(
	_request: AgencyListDomainStatusRequest
): ValidationError[] {
	return [];
}

export interface AgencyListDomainStatusItem {
	domain: string;
	status: AgencyDomainVerificationStatus;
	verification_token?: AgencyDomainVerificationToken;
	expires_at?: string; // ISO 8601 datetime
	last_verified_at?: string; // ISO 8601 datetime
	can_request_verification: boolean;
	last_attempted_at?: string; // ISO 8601 datetime — when verification was last requested
	next_verification_allowed_at?: string; // ISO 8601 datetime — earliest time next request is allowed (only when !can_request_verification)
}

export interface AgencyListDomainStatusResponse {
	items: AgencyListDomainStatusItem[];
	next_pagination_key?: string;
}
